cgi-bin/DW/Controller/Talk.pm

```diff
@@ -75,38 +85,66 @@ sub talkpost_do_handler {
     # input, since it may be not in UTF-8 in replies coming from mail
     # clients. We call it later.
 
+
     my $journalu = LJ::load_user($POST->{journal});
-    return error_ml('Unknown journal.  Please go back and try again.') unless $journalu;
+    return error_ml('Unknown journal.  Please go back and try again.') unless $journalu; # hmm, is error_ml right? -NF
+    $r->note( 'journalid', $journalu->userid ) if $r; # What the heck? -NF Oh, looks like maybe this gets looked up by S2.pm under some conditions.
+
+    my $entry = LJ::Entry->new( $journalu, ditemid => $POST->{itemid} + 0 );
+    unless ($entry) {
+        push @errors, 'talk.error.noentry';
+    }
+    my $talkurl = $entry->url;
 
     # libs for userpicselect
     LJ::Talk::init_iconbrowser_js()
         if $remote && $remote->can_use_userpic_select;
 
 
-    # show this error along with the regenerated comment form down below
+    # validate form auth (maybe)
     push @errors, LJ::Lang::ml('error.invalidform') if $remote && ! ( $skip_form_auth || LJ::check_form_auth($POST->{lj_form_auth} ) );
 
+    # validate the challenge/response value (anti-spammer)
+    my ($chrp_ok, $chrp_err) = LJ::Talk::validate_chrp1($POST->{'chrp1'});
+    unless ($chrp_ok) {
+        if ( $LJ::DEBUG{'talkspam'} ) {
+            my $ip = LJ::get_remote_ip();
+            my $ruser = $remote ? $remote->{user} : "[nonuser]";
+            carp( "talkhash error: from $ruser \@ $ip - $chrp_err - $talkurl\n" );
+        }
+        if ($LJ::REQUIRE_TALKHASH) {
```

Comment:
I also have been baking in most of these tunables to match production. Unless we need it for development, might as well simplify conditionals and remove a lot of these.